packages/kit-headless/src/components/tooltip/tooltip.test.ts

```diff
@@ -0,0 +1,240 @@
+import { expect, test, type Page } from '@playwright/test';
+import { createTooltipDriver } from './tooltip.driver';
+import { assertBoundingBoxExists } from '../../utils/test-utils';
+
+async function setup(page: Page, exampleName: string) {
+  await page.goto(`/headless/tooltip/${exampleName}`);
+
+  const driver = createTooltipDriver(page);
+
+  return {
+    driver,
+  };
+}
+
+test('@Visual diff', async ({ page }) => {
+  const { driver: d } = await setup(page, 'basic');
+  await expect(page).toHaveScreenshot('closed tooltip.png');
+
+  await d.getTrigger().hover();
+
+  await expect(page).toHaveScreenshot('opened tooltip.png');
+});
+
+test.describe('Mouse Behavior', () => {
+  test(`GIVEN a closed tooltip
+        WHEN hovering over the trigger
+        THEN the tooltip should open`, async ({ page }) => {
+    const { driver: d } = await setup(page, 'basic');
+    await expect(d.getTooltip()).toBeHidden();
+
+    await d.getTrigger().hover();
+
+    await expect(d.getTooltip()).toBeVisible();
+  });
+
+  test(`GIVEN an open tooltip
+        WHEN moving the mouse away from the trigger
+        THEN the tooltip should close`, async ({ page }) => {
+    const { driver: d } = await setup(page, 'basic');
+
+    await d.getTrigger().hover();
+    await expect(d.getTooltip()).toBeVisible();
+
+    await page.mouse.move(0, 0);
+    await expect(d.getTooltip()).toBeHidden();
+  });
+
+  test(`GIVEN an open tooltip
+        WHEN clicking on the trigger
+        THEN the tooltip should remain open`, async ({ page }) => {
+    const { driver: d } = await setup(page, 'basic');
+
+    await d.getTrigger().hover();
+    await expect(d.getTooltip()).toBeVisible();
+
+    await d.getTrigger().click();
+    await expect(d.getTooltip()).toBeVisible();
+  });
+});
+
+test.describe('Keyboard Behavior', () => {
+  test(`GIVEN an open tooltip
+        WHEN focusing on the trigger and pressing the 'Escape' key
+        THEN the tooltip should close`, async ({ page }) => {
+    const { driver: d } = await setup(page, 'basic');
+    await expect(d.getTooltip()).toBeHidden();
+
+    await d.getTrigger().focus();
+    await expect(d.getTooltip()).toBeVisible();
+
+    await d.getTrigger().press('Escape');
+    await expect(d.getTooltip()).not.toBeVisible();
+  });
+});
+
+test.describe('Placement and Positioning', () => {
+  const placements = ['top', 'right', 'bottom', 'left'] as const;
+
+  test.describe('Tooltip Placements', () => {
+    placements.forEach((placement) => {
```

Comment:
sometimes generated tests from a loop are harder to debug

An alternative approach is to encapsulate all of the setup in a dedicated setup function for this test (for example, everything until line 92 and just create this test 4 time and call the "custom setup" function with a param for the placement. 

It's not a must, but it might be easier to debug if we'll run into issues in the future.

Reply:
I refactored some helpers into the driver split out the different cases.